---
title: "Exploration"
output: 
  bookdown::html_document2:
    toc: true
    css: custom.css
bibliography: /Users/jhadfiel/Work/Tex/library/JarLib.bib
---
  
  <script src="hideOutput.js"></script>
  
  <script type="text/javascript">
  // When the document is fully rendered...
$(document).ready(function() {
  // ...select all header elements...
  $('h1, h2, h3, h4, h5').each(function() {
    // ...and add an id to them corresponding to their 'titles'
    $(this).attr('id', $(this).html());
  });
});
</script>

```{r setup_exploration, include=FALSE}

rm(list = ls())

knitr::opts_chunk$set(echo = TRUE)
run_model<-FALSE

#rmarkdown::render("/Users/jhadfiel/Work/Va_simulations/5_History_sim/6_Exploration.Rmd")

```

The following script is best ran in a conda environment with dependencies python=3.11.5 and numpy=1.26.0 (conda activate Va_anal).


# File Paths

Paths to various scripts that are used for running the simulations and extracting information from SLiM outputs. `base_path` specifies the path to all the scripts used for running simulations using SLiM and msprime - typically the "1_Simulations" folder in "Va_simulations". `analysis_path` specifies the path to the directory containing all scripts usefulf for analysing SLiM outputs: "Vw.Rmd", "Vw_sim_functions.Rmd", "2_Extract_mutations.py", and "3_Extract_genomes.py". `file_storage_path` is where the ?XXX files are stored? `slim_path` specifies the path to the folder of SLiM output.

```{r }
if(Sys.info()["nodename"]!="SCE-BIO-C06645"){message("WARNING: Command line arguments required!!!")}

### On AC3 ###

if(Sys.info()["nodename"]%in%c("bigfoot", "bigshot", "bigbird", "bigyin", "biggar", "bigwig", "c1", "c2", "c3", "c4", "c5", "c6")){
  
  base_path = "/ceph/users/marun/Va_simulations/1_Simulations" 
  analysis_path = "/ceph/users/marun/Va_simulations/2_Analysis"
  file_storage_path = "/data/obbard/Va_simulations/analyses" # File storage path is the designated storage space on AC3 instead of the /home directory on qm
  
}

### On Vera ###

if(Sys.info()["nodename"]=="vera.bio.ed.ac.uk"){
  
  base_path = "/data/home/msamant/Manas/Va_simulations/Github/Va_simulations/1_Simulations" 
  analysis_path = "/data/home/msamant/Manas/Va_simulations/Github/Va_simulations/2_Analysis"  
  file_storage_path = base_path
  
}

### On Eddie ###

if(grepl("ecdf.ed.ac.uk", Sys.info()["nodename"])){
  base_path = "/home/msamant/Va_simulations/1_Simulations"
  analysis_path = "/home/msamant/Va_simulations/2_Analysis"
  file_storage_path = "/exports/eddie/scratch/msamant"
}

### On Manas's PC

if(Sys.info()["nodename"]=="SCE-BIO-C06645"){
  
  if(Sys.info()["sysname"]=="Linux"){   ## Local Wsl
    
    base_path = "/mnt/c/Users/msamant/Documents/GitHub/Va_simulations/1_Simulations"
    analysis_path = "/mnt/c/Users/msamant/Documents/GitHub/Va_simulations/2_Analysis" 
    file_storage_path = base_path
    
  }else{                                ## Local windows
    
    base_path = "C:/Users/msamant/Documents/GitHub/Va_simulations/1_Simulations" 
    analysis_path = "C:/Users/msamant/Documents/GitHub/Va_simulations/2_Analysis" 
    file_storage_path = base_path
  }
  
}

### On Jarrod's PC ###

if(Sys.info()["nodename"]=="sce-bio-c04553"){  
  base_path="~/Work/Va_simulations/1_Simulations"
  analysis_path = "~/Work/Va_simulations/2_Analysis"
  file_storage_path = base_path
  slim_path = "/Volumes/hadfield/Va_simulations/sim_files"
}else{
  slim_path = "/mnt/u/Datastore/CSCE/biology/groups/hadfield/Va_simulations/sim_files"
}


extract_genomes_path = file.path(analysis_path, "3_Extract_genomes.py")                                           ## Python script extracting mutations and genomes from the SLiM output file 
extract_mut_path = file.path(analysis_path, "2_Extract_mutations.py")                                             ## Python script extracting mutations from the SliM output file

slim_output_path = file.path(slim_path, "SLiM_outputs")
# Path to SLiM output              
slim_param_path = file.path(slim_path, "sim_params") 
# Path to SLiM simulation parameters

temp_files_path = file.path(file_storage_path, "temp_files")  
```

# Load packages and functions

```{r packages}
if(Sys.info()["nodename"]=="bigyin"){stop("Bigyin cannot run asreml-r. Use a different code.")}

library(MCMCglmm)
library(asreml)
library(Matrix)
#library(rmutil)
library(pryr) ## For tracking memory usage using mem_used()
library(bigalgebra)
library(RhpcBLASctl)

  
#################################
#### Load Jarrod's functions ####
#################################

functions_only=TRUE ## Read only the functions

rmarkdown::render(file.path(analysis_path, "Vw.Rmd"))

##############################
### Load Manas's functions ###
##############################

rmarkdown::render(file.path(analysis_path, "Vw_sim_functions.Rmd"))
```

# Get data

First enter the Set_ID of the simulations to be analysed

```{r }
Set_ID = "bigfoot_2024-08-01_13_03_32.727659_9.64489795918367e-07_1.4_1.4_1000_10_3_estimate_0"
```

and then extract data

```{r get_data}

sim_data = extract_slim_data(Set_ID = Set_ID,
                             sim = 1,
                             slim_output_path = slim_output_path, 
                             sim_param_path = slim_param_path,
                             extract_genomes_path = extract_genomes_path, 
                             extract_mut_path = extract_mut_path,
                             mutations_path = temp_files_path, 
                             c_matrix_path = temp_files_path, 
                             randomise = TRUE)
```

# Analyse data

```{r analyse_data}
parents_info = analyse_parents(c0 = sim_data$c0,  
                               list_alpha = sim_data$list_alpha,             
                               LDdelta=FALSE,         
                               SNPs = sim_data$SNPs,                   
                               RecombRate = sim_data$sim_params$r_expt,             
                               HapLength = sim_data$sim_params$sequence_length,              
                               AtleastOneRecomb=FALSE)
```

# Model parameters

```{r model_parameters}
# Analysis parameters
proj="BLoM" # projection type for allele frequencies: "LoM", "BLoM", "L" or "N"
LDdelta = FALSE
pa = 1
Vs = "LoNL" # "L" or "LoNL"
method="REML" # Can be "REML" or "MCMC"
randomise = TRUE # Should the reference allele be randomised for analysis?
bigalgebra = FALSE # Should bigalgebra be used for eigendecomposition?

# How is pdelta to be estimated? 
# Can be "optim" (using the function optim()), or "fixed" or "manual"(estimated by manually scanning a range of pdelta values)

pdelta_method = "optim" # "optim" or "manual" or "fixed" or "no_analysis". If this is "no_analysis", the estimate of Vw is not calculated, but the rest of the code still runs.

if(pdelta_method=="fixed"){
  pdelta = 0 # Can be specified to any value
}

if(pdelta_method=="optim"){
  pdelta = NA # This triggers the use of optim() inside the function Vw_model()
}

if(pdelta_method=="manual"){
  
  nseq<-20 # The number of times pdelta is to be varied 
  pdelta_l = -1 # Lower limit of pdelta
  pdelta_u = 1 # Upper limit of pdelta
  pdelta<-seq(pdelta_l, pdelta_u, length=nseq)
  
}

# How should bdelta[1] (intercept) and bdelta[2] (slope of (p-q)) be estimated

bdelta_method = "estimate"  # Can be "fixed" or "estimate"

if(bdelta_method=="estimate"){
  bdelta = c(NA, NA)
}else{
  bdelta = c(0, 0) # This only estimates the intercept while keeping the slope fixed at 0
}
```
Fit model

```{r fit_model, eval=run_model}

m1<-Vw_model(C0 = sim_data$c0,          # parental genotypes (rows individuals, columns loci, coded as 0, 1/2 or 1) 
             nR = parents_info$nR,          # matrix of non-recombinant probabilities between loci
             pbar1 = sim_data$pbar1,       # vector of allele frequencies at time-point 1
             ngen1=1,     # number of generations between parents and time-point 1
             pbar2 = sim_data$pbar2,       # vector of allele frequencies at time-point 2
             ngen2 = (sim_data$sim_params$ngen_expt) + 1,       # number of generations between parents and time-point 2
             nind = sim_data$sim_params$n_ind_exp,        # population size in each replicate
             proj=proj, # projection type for allele frequencies: "LoM", "BLoM", "L" or "N"
             LDdelta = LDdelta,
             pa = pa,
             pdelta = pdelta,
             bdelta = bdelta,
             Vs = Vs,
             method = method,
             L = parents_info$L,    # list with elements UL and DL
             svdL = NULL,    # list with elements UL and DL
             bigalgebra = bigalgebra, 
             tol = sqrt(.Machine$double.eps))
```

# Extract estimates

```{r extract_estimates, eval=run_model}
vA_est = m1$Vw_est # Store the estimate from the model with the highest log likelihood
pdelta_est = m1$pdelta # The sample() functions ensures that only one value is selected in case there are multiple points with the highest LL
pdelta_var_est = m1$pdelta_var
bdelta_intercept_est = m1$bdelta[1]
bdelta_slope_est = m1$bdelta[2]
bdelta_var_est = paste(m1$bdelta_var[1,1], m1$bdelta_var[2,2], m1$bdelta_var[1,2], sep = "_")
sigma2delta_est = summary(m1$model)$varcomp[1,1]
```


