---
title: "Vw_sim_functions"
output: html_document
date: "2024-07-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Overview

Here we describe the functions that can be used to extract useful data from the outputs of SLiM simulations and then analyse it to obtain estimates of Vw. 

### Function to extract SLiM data

This function reads the SLiM output produced in the simulation, extracts the useful bits from those files, optionally randomises reference alleles, and returns -

1. c0 matrix in the parents' generation (c0)
2. Vector of alphas in the parents generation (list_alpha)
3. Vector of positions of mutations (SNPs)
4. pbar1 
5. pbar2
6. A dataframe containing parameters used in the simulations (sim_params)

It depends on the python scripts 3_Extract_genomes.py and 2_Extract_mutations.py

```{r extract_slim_data}

extract_slim_data = function(Set_ID,                # The unique ID of the set of simulations that are controlled by a single R script
                             sim = 1,               # Each set can have multiple sims, but - on the cluster sim must always 1
                             unzip = FALSE,          # Should the SLiM output file be unzipped, read, and then zipped back?
                             slim_output_path,      # The directory where the SLiM outputs (for parents and experimental replicates) are stored (as .txt files)
                             sim_param_path,        # The path to the directory where the .csv file containing simulation parameters is stored
                             extract_genomes_path,  # The path to the python script that extracts genomes and mutations from SLim outputs (3_Extract_genomes.py)
                             extract_mut_path,      # The path to the python script that extracts mutations from SLim outputs (2_Extract_mulations.py)
                             mutations_path,        # The directory where extracted mutations are to be stored (temp files)
                             c_matrix_path,         # The directory where extracted genomes are to be stored (temp files)
                             n_sample=NULL,         # Number of individuals sampled from the parents' generation (useful if n_ind_exp is large)
                             randomise = TRUE,      # Optionally the reference allele can be randomised
                             delete_temp_files = TRUE, 
                             verbose=TRUE){     
  
  ################################
  ## Read simulation parameters ##
  ################################
      
  sim_params = read.csv(paste(sim_param_path, "/", Set_ID, "_Data.csv", sep = ""), header = T)
  # Restrict the data to current sim if the data frame has data from multiple sims
  if("sim"%in%colnames(sim_params)){sim_params = sim_params[sim_params$sim==sim,]}
  
  # Load sim parameters
  
  n_ind_exp = sim_params$n_ind_exp     # Number of individuals in the experiment
  n_cages = sim_params$n_cages         # Number of replicates in the experiment
  ngen_expt = sim_params$ngen_expt     # Number of generations over which allele frequency changes are calculated
  end_gen = sim_params$end_gen         # The generation number of the parents' generation

  
  # If n_sample is not provided extract the genomes of all individuals in the parents' generation
  
  if(is.null(n_sample)){n_sample = n_ind_exp}
  if(n_sample>n_ind_exp){stop("n_sample cannot be greater than n_ind_exp")}
  
  if(verbose){
    message("Reading the state of the population in the parent's generation...")
    message("Extracting mutations and genomes from the parents' generation...")
  }
  
  # Unzip the SLiM output file
  
  if(unzip){
    if(verbose)message("Unzipping the SLiM output file...")
    system(paste("gunzip", paste(slim_output_path, "/", Set_ID, "_sim", sim, "_output_parents.txt.gz", sep = "")))
  }

  system(paste("python", 
               extract_genomes_path,                        # Path of the python script (3_Extract_genomes.py)
               paste(slim_output_path, "/", Set_ID, "_sim", sim, "_output_parents.txt", sep = ""),  # Path of the .txt file containing the SLiM output for the parent's generation 
               paste(mutations_path, "/", Set_ID, "_sim", sim, "_mutations_parents.txt", sep = ""), # Path of the .txt output file containing the mutations in the parents' generation 
               paste(c_matrix_path,"/", Set_ID, "_sim", sim, "_c_matrix_parents.csv", sep = ""),    # Path of the .csv output file containing the c matrix for genomes in the parents' generation
               n_sample))                                                                 # Number of individuals to be sampled randomly to construct the c matrix (just for space issues in case n_ind_exp is very large). Typically should be set to same as n_ind_exp
  
  # Rezip the SLiM output file
  
  if(unzip){
    if(verbose)message("Re-zipping the unzipped SLiM output file...")
    system(paste("gzip", paste(slim_output_path, "/", Set_ID, "_sim", sim, "_output_parents.txt", sep = "")))
  }

  
  # Read genomes
  c_genome = read.csv(paste(c_matrix_path,  "/", Set_ID, "_sim", sim, "_c_matrix_parents.csv", sep =""), header=F) # as.integer done to avoid scientific notation
  
  # Delete the .csv file
  if(delete_temp_files){system(paste("rm", paste(c_matrix_path,  "/", Set_ID, "_sim", sim, "_c_matrix_parents.csv", sep =""), sep = " "))}
  
  c_genome = as.matrix(c_genome)
  
  # Convert genome data into individual data (rows are individuals and columns are allele counts {0,1 or 2} at various sites). 
  # Note that genome 1 and genome 2 are from individual 1; 3 and 4 are from individual 2, and so on
  
  if(verbose){
    message("Calculating C0...")
  }

  n_individuals = nrow(c_genome)/2
  n_sites = ncol(c_genome)
  
  c_ind = c_genome[seq(1, 2*n_individuals, 2),] + c_genome[seq(2, 2*n_individuals, 2),]
  
  # Remove c_genome to save memory
  rm("c_genome")
  gc(verbose = FALSE)
  
  # If one samples individuals from the parents' generation while building the c matrix (i.e. when sample_size is less than n_ind_exp), the sample may not contain some low frequency mutations, i.e. some loci are not segregating in the sample, but are in the parents' population
  
  # identify the loci that are missing in the sample
  
  missing_loci = which(colSums(c_ind)==0)
  retained_loci = which(colSums(c_ind)!=0)
  
  # Trim the c matrix to contain only the retained loci, i.e. the loci that are segregating in the sample
  
  c_ind = c_ind[,retained_loci]
  

  ## Read the list of mutations output generated by SLiM and subsequently cropped out as a separate file by Python (for the parents' generation)
  
  mutations_0 = read.table(paste(mutations_path, "/", Set_ID, "_sim", sim, "_mutations_parents.txt", sep = ""), header=T, sep = " ")
  
  # Delete the .txt file
  if(delete_temp_files){system(paste("rm ", mutations_path, "/", Set_ID, "_sim", sim, "_mutations_parents.txt", sep = ""))}
  
  # Sort mutations based on the Temp_ID, so that the order of loci matches the order in c_ind
  
  mutations_0 = mutations_0[order(mutations_0$Temp_ID),]
  
  # Trim mutations to contain only the retained loci, i.e. the loci that are segregating in the sample
  
  mutations_0 = mutations_0[retained_loci,]
  
  list_alpha = 2*(mutations_0$s)   # Vector of alphas
  SNPs = mutations_0$Position      # Vector of positions of mutations in the parents
  
  ############################################################################
  ############################################################################
  ####### Calculate allele frequencies in the experiment for each cage #######
  ############################################################################
  ############################################################################  
  
  # Create empty vector to create the data frame containing the following variables as columns:
  # 1. Raw delta P
  # 2. Projected delta P
  # 3. Cage ID (replicate)
  # 4. Locus ID
  
  d_proj = c()
  d_raw = c()
  P_matrix = c()  
  
  for (cage in (1:n_cages)){
    
    ###############################
    ###### extract mutations ######
    ###############################
    
    
    # There are two command line arguments (1. Path of the SLiM output file, 2. Path where mutations are to be written) 
    
    if(verbose){
      message(paste("Extracting mutations and storing allele frequencies in cage ", cage, " of simulation ", sim, "...", sep = ""))
    }

    for (gen in (end_gen + 1):(end_gen + 1 + ngen_expt)){
      
      # Unzip SLiM output files
      
      if(unzip){
        if(verbose){message("Unzipping the SLiM output file...")}
        system(paste("gunzip", paste(slim_output_path, "/", Set_ID, "_sim", sim, "_cage", cage, "_output_experiment_", as.integer(gen), ".txt", sep = "")))
      }
      
      system(paste("python", 
                   extract_mut_path, 
                   paste(slim_output_path, "/", Set_ID, "_sim", sim, "_cage", cage, "_output_experiment_", as.integer(gen), ".txt", sep = ""), 
                   paste(mutations_path, "/", Set_ID, "_sim", sim, "_cage", cage, "_mutations_", as.integer(gen), ".txt", sep = ""))) # as.integer done to avoid scientific notation
      
      # Re-zip unzipped SLiM files
      if(unzip){
        if(verbose){message("Re-zipping the unzipped SLiM output file...")}
        system(paste("gzip", paste(slim_output_path, "/", Set_ID, "_sim", sim, "_cage", cage, "_output_experiment_", as.integer(gen), ".txt", sep = "")))
      }
    }
    
    
    ### Create an empty matrix to store allelic frequencies in each generation
    
    # If some of the loci get fixed/lost in subsequent generations, NAs should be inserted
    
    # Create an empty vector to store allele frequencies
    
    P = c()
    
    # store the frequencies in the parent's generation in P
    # Frequency = (Number of genomes)/(2*popsize)
    
    P = cbind(P, mutations_0$Number/(2*n_ind_exp))
    
    ### Loop through the generations of the experiment identifying mutations that were present in the parents' generation (using permanent IDs) and recording their frequencies
    
    for (gen in (end_gen+1):(end_gen+ngen_expt+1)){
      
      # Read the file storing mutation information
      mut = read.csv(paste(mutations_path, "/", Set_ID, "_sim", sim, "_cage", cage, "_mutations_", as.integer(gen), ".txt", sep = ""), sep = " ") # as.integer gets rid of scientific notation
      
      # Delete the .txt file
      if(delete_temp_files){system(paste("rm ", mutations_path, "/", Set_ID, "_sim", sim, "_cage", cage, "_mutations_", as.integer(gen), ".txt", sep = ""))}
      
      # Sort mutations based on the Temp_ID, so that the order of loci matches the order in the C and L matrices
      
      mut = mut[order(mut$Temp_ID),]
      
      # Create an empty vector to store frequencies of mutations in the current generation
      freq = c()
      
      # Loop through the permanent IDs of  mutations segregating in end_gen (parents' generation)
      # i.e. Loop through Permanent IDs in mutations_0
      # Check if each mutation is present in the current generation
      # If present, record the frequency in freq, otherwise add either 0 or 1 to freq using the round() function
      
      for(mutation in mutations_0$Permanent_ID){
        if(mutation %in% mut$Permanent_ID){freq = c(freq, (mut[which(mut$Permanent_ID==mutation),]$Number)/(2*n_ind_exp))
        }else{
          freq = c(freq, round(P[which(mutations_0$Permanent_ID==mutation), gen - end_gen]))
          
        }
      }
      
      # Add the vector freq to P
      
      P = cbind(P, freq)
      
    }
   
    P_matrix = rbind(P_matrix, P)
    
  }
  
  
  ##### Calculate matrices of allele frequencies with rows as replicates ######
  
  pbar1 = matrix(NA, nrow = n_cages, ncol = n_sites)
  pbar2 = matrix(NA, nrow = n_cages, ncol = n_sites)
  for (i in 1:n_cages){
    pbar1[i,] = P_matrix[((i-1)*n_sites + 1):((i-1)*n_sites + n_sites),2] ## Matrix of frequencies at the start of the experiment (ie F1 generation)
  }
  
  for (i in 1:n_cages){
    pbar2[i,] = P_matrix[((i-1)*n_sites + 1):((i-1)*n_sites + n_sites),ngen_expt + 2] ## Matrix of frequencies at the end of the experiment
    
    
  }
    
  if(randomise){
    
      #########################################################
      ######## Randomise the reference allele in c_ind ########
      #########################################################
      
      # Randomly change the reference allele
      # This can be done as follows:
      # Add 0s to the allele counts of those alleles that stay the same, and -2 to those alleles that are to be switched
      # Then take a mod
      # Remember c_ind is a matrix of 0s 1s and 2s
      
      if(verbose){
        message("Randomising reference alleles in c_ind...")
      }
        
      # Generate a random vector of 0s (for no change) and -1s (for loci where the reference allele is to be switched)
      ran_vect = sample(c(0, -1), ncol(c_ind),  replace = T) 
      
      # Create a matrix with with as many rows as c_ind. 
      # Each row of this matrix should be made up of two times ran_vect (since we are working with allele counts, not frequencies). 
      # Because the same changes need to be applied to each individual
      
      ran_matrix = t(matrix(2*ran_vect, nrow = ncol(c_ind), ncol = nrow(c_ind)))
      
      # Calculate the allele counts of the new (randomised) reference alleles
      
      c_ind = abs(c_ind + ran_matrix)
      
      ####################################################################
      ######## Randomise the reference allele in L and list_alpha ########
      ####################################################################
      
      if(verbose){
        message("Randomising reference alleles in list_alpha...")
      }

      # Create a new vector of -1s (wherever ran_vect has -1) and 1s (wherever ran_vect has 0)
      ran_vect_alpha = ifelse(ran_vect==-1, -1, 1)
      
      list_alpha = list_alpha*ran_vect_alpha
      
      
      ###########################################################################
      ######## Randomise the reference allele in p_parents, pbar1 and pbar2 #####
      ###########################################################################
      
      # Randomly change the reference allele
      # This can be done as follows:
      # Add 0s to the frequencies of those alleles that stay the same, and -1 to those alleles that are to be switched
      # Then take a mod
      
      if(verbose){
        message("Randomising reference alleles in pbar1 and pbar2...")
      }

      # Create a matrix with with as many rows as pbar1. Each row of this matrix should be made up of ran_vect (computed while randomising c_ind). 
      # Because the same changes need to be applied to each cage
      
      ran_matrix_pbar = t(matrix(ran_vect, nrow = ncol(pbar1), ncol = nrow(pbar1)))
      
      # Calculate the frequencies of the new (randomised) reference alleles
      
      pbar1 = abs(pbar1 + ran_matrix_pbar)
      pbar2 = abs(pbar2 + ran_matrix_pbar)
      
    
  }
 
  return(list(c0=c_ind/2, list_alpha=list_alpha, SNPs=SNPs, pbar1=pbar1, pbar2=pbar2, sim_params=sim_params[,1:19]))
    
}
```

### Function to analyse the parents' generation

This function uses data from the parents' generation to calculate:
(a) Two matrices (L and nR)
(b) A host of scalar parameters of interest in the parents' generation (eg. true vA, true va, number of segregating sites for each type of mutations, properties of the distribution of alphas, etc.)

It depends on form_nR() and alpha_distribution().

```{r analyse_parents}

analyse_parents = function(c0,  
                           list_alpha,             # Vector of alphas
                           LDdelta=FALSE,          # Whether L or diag(L) should be used to model vA using alpha distribution
                           SNPs,                   # Vector of positions of SNPs
                           RecombRate,             # Recombination rate between adjacent sites
                           HapLength,              # Genome size
                           tol=sqrt(.Machine$double.eps),
                           AtleastOneRecomb=FALSE, # Should there be at least one cross-over per meiosis?
                           calc_nR=TRUE, 
                           verbose=TRUE){ 
  
  ########################
  ## Calculate L and nR ##
  ########################
  
  n_individuals = nrow(c0)

  if(verbose){
    message("Calculating L...")
  }
  L = cov(c0)*(n_individuals-1)/(n_individuals) 
  
  if(LDdelta){
    if(verbose){
      message("Performing SVD on c0...")
    }  
    svdC<-svd(scale(sqrt(1/n_individuals)*c0), nu = 0)
    retain<-sum(svdC$d>tol)
    UL<-svdC$v[,1:retain]
    DL<-svdC$d[1:retain]
  }

  if(calc_nR){
  
    if(verbose){
      message("Calculating the matrix of non-recombinant fractions...")
    }

    nR = form_nR(SNPs, RecombRate, HapLength, AtleastOneRecomb)
  }else{
    nR = NULL
  }

  # Segregating sites
  
  seg_sites = length(list_alpha)
  seg_sites_neu = sum(list_alpha==0)
  seg_sites_ben = sum(list_alpha>0)
  seg_sites_del = sum(list_alpha<0)
  
  ##############################################
  ## True additive genic and genetic variance ##
  ##############################################
  
  pbar0 = colMeans(c0) 
  diversity = pbar0*(1 - pbar0)/2
  mean_diversity = mean(diversity)

  if(verbose){
    message("Calculating the true additive genetic and genic variances...")
  }  
  va_true = sum(diversity*list_alpha^2)          # additive genic variance
  vA_true = t(list_alpha)%*%L%*%list_alpha       # Additive genetic variance
  
  ###################################################################
  ##### Calculate empirical properties of distribution of alphas ####
  ###################################################################
  
  if(verbose){
    message("Calculating properties of the empirical distribution of alphas...")
  }  
  alpha_properties = alpha_distribution(alpha = list_alpha, p = pbar0)
  
  pdelta_emp = alpha_properties$pdelta
  bdelta_intercept_emp = alpha_properties$bdelta_0
  bdelta_slope_emp = alpha_properties$bdelta_1
  sigma2delta_emp = alpha_properties$sigma2delta

  # Calculate vA from these empirical properties of alpha
      
  bdelta_emp = c(bdelta_intercept_emp, bdelta_slope_emp)
  
  if(LDdelta){
    TrV<-sum(DL^(2*(pdelta_emp+1)))*sigma2delta_emp
    aLa<-t((alpha_properties$X)%*%bdelta_emp)%*%L%*%(alpha_properties$X)%*%bdelta_emp-sum(diag(t(alpha_properties$X)%*%L%*%(alpha_properties$X)%*%(alpha_properties$S)))
  }else{
    TrV<-sum(diag(L)^(pdelta_emp+1))*sigma2delta_emp
    aLa<-t((alpha_properties$X)%*%bdelta_emp)%*%L%*%(alpha_properties$X)%*%bdelta_emp-sum(diag(t(alpha_properties$X)%*%L%*%(alpha_properties$X)%*%(alpha_properties$S)))
  }
  
  vA_alpha_emp<-TrV+aLa
  
  return(list(L=L, nR=nR, svdL=if(LDdelta){list(UL=UL, DL=DL)}else{NULL}, seg_sites=seg_sites, seg_sites_neu=seg_sites_neu, seg_sites_ben=seg_sites_ben, seg_sites_del=seg_sites_del, mean_diversity=mean_diversity, va_true=va_true, vA_true=vA_true, vA_alpha_emp=vA_alpha_emp, parameters=alpha_properties[1:5]))

}
```

### Function to peform analyses on SLiM outputs

This function takes in Set_ID of the SLiM simulation and a host of analysis parameters as the input, and performs the following steps -

(a) Extract simulation data using extract_slim_data()
(b) Extract useful information about the parents' generation using analyse_parents()
(c) Fit models and perform analyses using Vw_model()
(d) Return all simulation and analysis parameters as well as outputs of analyses 
(e) Store simulation and analysis parameters as well as outputs of analyses as a .csv file

```{r analyse_sim}

analyse_sim = function(Set_ID,                # The unique ID of the set of simulations that are controlled by a single R script
                       sim = 1,               # Each set can have multiple sims, but - on the cluster sim must always 1
                       unzip = FALSE,          # Should the SLiM output file be unzipped, read, and then zipped back?
                       slim_output_path,      # The directory where the SLiM outputs (for parents and experimental replicates) are stored (as .txt files)
                       sim_param_path,        # The path to the directory where the .csv file containing simulation parameters is stored
                       extract_genomes_path,  # The path to the python script that extracts genomes and mutations from SLim outputs
                       extract_mut_path,      # The path to the python script that extracts mutations from SLim outputs
                       mutations_path,        # The directory where extracted mutations are to be stored (temp files)
                       c_matrix_path,         # The directory where extracted genomes are to be stored (temp files)
                       output_path,           # The path where the final data file is to be stored
                       n_sample=NULL,         # Number of individuals sampled from the parents' generation (useful if n_ind_exp is large)
                       randomise = TRUE,      # Optionally the reference allele can be randomised
                       delete_temp_files = TRUE,
                       proj = "BLoM",           # projection type for allele frequencies: "LoM", "BLoM", "L" or "N"
                       LDdelta = FALSE,       # Should L or diag(L) be considered while modelling distribution of alphas
                       pa = 1,
                       Vs = "LoNL",           # "L" or "LoNL"
                       method="REML",         # Can be "REML" or "MCMC"
                       pdelta = NA,           # If NA pdelta is estimated using optim()
                       bdelta = c(NA, NA),    # If c(NA,NA) both bedelta intercept and slope are estimated
                       AtleastOneRecomb=FALSE,
                       verbose = TRUE
                       ){
  
  ### Extract sim data ###
  
  if(verbose){message("Extracting simulation data...")}

  sim_data = extract_slim_data(Set_ID = Set_ID,
                               sim = sim,
                               unzip = unzip,
                               slim_output_path = slim_output_path, 
                               sim_param_path = sim_param_path,
                               extract_genomes_path = extract_genomes_path, 
                               extract_mut_path = extract_mut_path,
                               mutations_path = mutations_path, 
                               c_matrix_path = c_matrix_path, 
                               randomise = randomise)
  ### Analyse parents ###
  
  if(verbose){message("Analysing the parents' generation...")}

  parents_info = analyse_parents(c0 = sim_data$c0,  
                                 list_alpha = sim_data$list_alpha,             
                                 LDdelta=LDdelta,         
                                 SNPs = sim_data$SNPs,                   
                                 RecombRate = sim_data$sim_params$r_expt,             
                                 HapLength = sim_data$sim_params$sequence_length,              
                                 AtleastOneRecomb=AtleastOneRecomb)

  ### Fit model ###
  
  if(verbose){message("Performing analyses...")}

  m1<-Vw_model(C0 = sim_data$c0,          
               nR = parents_info$nR,             
               pbar1 = sim_data$pbar1,      
               ngen1=1,     
               pbar2 = sim_data$pbar2,       
               ngen2 = (sim_data$sim_params$ngen_expt) + 1,       
               nind = sim_data$sim_params$n_ind_exp,        
               proj=proj,
               LDdelta = LDdelta,
               pa = pa,
               pdelta = pdelta,
               bdelta = bdelta,
               Vs = Vs,
               method = method,
               L = parents_info$L,    
               svdL = parents_info$svdL,           # list with elements UL and DL
               tol = sqrt(.Machine$double.eps))


  vA_est = m1$Vw_est 
  pdelta_est = m1$pdelta 
  pdelta_var_est = m1$pdelta_var
  bdelta_intercept_est = m1$bdelta[1]
  bdelta_slope_est = m1$bdelta[2]
  bdelta_var_est = paste(m1$bdelta_var[1,1], m1$bdelta_var[2,2], m1$bdelta_var[1,2], sep = "_")
  sigma2delta_est = summary(m1$model)$varcomp[1,1]
    
  ### Save file ###
  
  if(verbose){message("Saving data...")}
 
  sim_params = sim_data$sim_params

  analysis_data = data.frame("proj"=proj, "LDdelta"=LDdelta, "pa"=pa, "Vs"=Vs, "randomise"=randomise, "pdelta_method"=pdelta, "bdelta_method"=paste(bdelta[1], bdelta[2], sep="_"), "va_true"=parents_info$va_true, "vA_true"=parents_info$vA_true, "vA_est"=vA_est, "vA_alpha_emp"=parents_info$vA_alpha_emp, "pdelta_emp"=parents_info$parameters$pdelta, "bdelta_intercept_emp"=parents_info$parameters$bdelta_0, "bdelta_slope_emp"=parents_info$parameters$bdelta_1, "sigma2delta_emp"=parents_info$parameters$sigma2delta, "pdelta_est"=pdelta_est, "pdelta_var_est"=pdelta_var_est, "bdelta_intercept_est"=bdelta_intercept_est, "bdelta_slope_est"=bdelta_slope_est, "bdelta_var_est"=bdelta_var_est, "sigma2delta_est"=sigma2delta_est, "seg_sites"=parents_info$seg_sites, "seg_sites_neu"=parents_info$seg_sites_neu, "seg_sites_ben"=parents_info$seg_sites_ben, "seg_sites_del"=parents_info$seg_sites_del, "mean_diversity"=parents_info$mean_diversity)

  analysis_data = cbind(sim_params, analysis_data)
  write.table(rbind(names(analysis_data), analysis_data), file = paste(output_path, "/", Set_ID, "_sim_", sim, "_Data_analysis.csv", sep = ""),col.names = FALSE, row.names = FALSE, sep = ",")
  
  return(analysis_data)
  
}

```

