rm(list = ls())

# source("C:/Users/msamant/Documents/GitHub/Va_simulations/1_Simulations/00_Control_sims.R")               ## Local windows

# source("/mnt/c/Users/msamant/Documents/GitHub/Va_simulations/1_Simulations/00_Control_sims.R")           ## Local Wsl
# Rscript /mnt/c/Users/msamant/Documents/GitHub/Va_simulations/1_Simulations/00_Control_sims.R

# source("/data/home/msamant/Manas/Va_simulations/Github/Va_simulations/1_Simulations/00_Control_sims.R")  ## ON VERA

# source("/ceph/users/marun/Va_simulations/1_Simulations/00_Control_sims.R")                               ## ON QMASTER


##########################################
######## Code parallelization ############
##########################################

# parallel -j 18 --colsep "\s" --delay 5.0 Rscript 00_Control_sims.R :::: 000_parameter_grid.txt
# https://www.danielecook.com/using-gnu-parallel-for-bioinformatics/

########################################################################
########### paths of various scripts and functions #####################
########################################################################

if(Sys.info()["nodename"]!="SCE-BIO-C06645"){message("WARNING: Command line arguments required!!!")}

### Base path (depending on the system) ###

### On AC3 ###

if(Sys.info()["nodename"]%in%c("bigfoot", "bigshot", "bigbird", "bigyin", "biggar", "bigwig", "c1", "c2", "c3", "c4", "c5", "c6")){
  
  base_path = "/ceph/users/marun/Va_simulations/1_Simulations" # Path to all the scripts except "Vw.Rmd"
  file_storage_path = "/data/obbard/Va_simulations/analyses" # File storage path is the designated storage space on AC3 instead of the /home directory on qm
  
}
  
### On Vera ###
  
if(Sys.info()["nodename"]=="vera.bio.ed.ac.uk"){
  
  base_path = "/data/home/msamant/Manas/Va_simulations/Github/Va_simulations/1_Simulations" ## ON VERA
  file_storage_path = base_path
  
}

### On Eddie ###

if(grepl("ecdf.ed.ac.uk", Sys.info()["nodename"])){
  base_path = "/home/msamant/Va_simulations/1_Simulations"
  file_storage_path = "/exports/eddie/scratch/msamant"
}

### On Manas's PC
    
if(Sys.info()["nodename"]=="SCE-BIO-C06645"){

  if(Sys.info()["sysname"]=="Linux"){
    
    base_path = "/mnt/c/Users/msamant/Documents/GitHub/Va_simulations/1_Simulations" ## Local Wsl
    file_storage_path = base_path
    
  }else{
    
    base_path = "C:/Users/msamant/Documents/GitHub/Va_simulations/1_Simulations" ## Local windows
    file_storage_path = base_path
  }

}
  

# Paths to various scripts that are used for running the simulations and extracting information from SLiM outputs

msprime_burnin_path = file.path(base_path, "0_neutral_burnin.py")                                             ## msprime script generating the initial sequences
slim_history_path = file.path(base_path, "1_History.slim")                                                    ## SLiM script running the history
msprime_add_neutral_path = file.path(base_path, "2_History_add_neutral_mut.py")                               ## msprime script that adds neutral mutations to the tree sequence generated by 1_History.slim
slim_flip_sel_coef_path = file.path(base_path, "3_Change_sel_coef.slim")                                     ## SLiM script optionally flipping selection coef
slim_expt_path = file.path(base_path, "4_Experiment.slim")                                                    ## SLiM script running the experiment

# Paths to various directories

slim_output_path = paste(file_storage_path, "/b_Interim_files/SLiM_outputs", sep = "")                        ## Path where SLiM and msprime output files are stored
output_path = paste(file_storage_path, "/c_Output", sep = "")                                                 ## Path where .csv file(s) containing simulation parameters are to be stored 


###################################
###### Create file structure ######
###################################

##### Create directory that stores outputs

system(paste("mkdir -p", output_path)) # Make directory but ignore if already present

##### Create directory that stores temp files and directories therein

system(paste("mkdir -p", paste(file_storage_path, "/b_Interim_files", sep = "")))
system(paste("mkdir -p", paste(file_storage_path, "/b_Interim_files/SLiM_outputs", sep = "")))

###################################################
#### Command line arguments fed to this script ####
###################################################

# The ith command line arguments can be retrieved with commandArgs(trailingOnly = TRUE)[i]
# Command line arguments are not used if running on Manas's PC

# Arg 1 = mu
# Arg 2 = r*sequence_length
# Arg 3 = r_expt*sequence_length
# Arg 4 = n_ind_exp
# Arg 5 = n_cages
# Arg 6 = ngen_expt
# Arg 7 = flip_sel_coef
# Arg 8 = mut_ratio

#####################################################
############ Simulation Parameters ##################
#####################################################

if(Sys.info()["nodename"]!="SCE-BIO-C06645"){
  # Print command line arguments to screen
  print(c("Mutation rate", "recombination rate (M)", "Population size", "No. of cages", "Experimental generations" ))
  print(commandArgs(trailingOnly = TRUE))
}


record = TRUE                          # Should the data of the simulations be recorded in a .csv file 
trim_exp_files = FALSE                 # Should the SLiM output files for the experiment be trimmed ti include just the information on mutations to save space.
del_files = TRUE                       # Should the .trees files be deleted at the end to save space?
compress_files = FALSE                 # Should .txt and .trees files be compressed using gzip

Job_ID = "zero_test"            # Job ID will be prefixed to Set_IDs so that output files can be more easily parsed

nsims = 1                              # Number of simulations - MUST be 1 if running on a cluster

n_cages = ifelse(Sys.info()["nodename"]=="SCE-BIO-C06645", 10, (as.numeric(commandArgs(trailingOnly = TRUE)[5])))     # The number of replicate cages in the experiment

end_gen = 2                        # How many generations should the SLiM simulation run for while simulating the history (burnin) (for sims without burnin this has to be 2)

if(end_gen<2){stop("end_gen must be an integer greater than or equal to 2")}

output_freq = 1000                     # The frequency with which SLiM outputs are to be generated for the analysis of history 
ngen1 = 1                              # How many generations between the 1st expt generatio and the parents
ngen2 = ifelse(Sys.info()["nodename"]=="SCE-BIO-C06645", 4, (as.numeric(commandArgs(trailingOnly = TRUE)[6])))       # How many generations between the last expt generatio and the parents                        # How many generations should allele frequency changes be calculated over in the experiment
flip_sel_coef = ifelse(Sys.info()["nodename"]=="SCE-BIO-C06645", 0, as.numeric(commandArgs(trailingOnly = TRUE)[7])) # (1 for TRUE and 0 for FALSE) Multiply the selection coefficients in the parents' generation by -1 or 1 randomly (for testing purposes)

###########################################
########### Pop gen parameters ############
###########################################

Ne = 1.33e+06                          # Effective population size
n_ind = 2500                           # Number of individuals to be sampled in msprime and then run forward in SLiM
n_ind_exp = ifelse(Sys.info()["nodename"]=="SCE-BIO-C06645", 1000, (as.numeric(commandArgs(trailingOnly = TRUE)[4])))                       # The population size of the experiment. In 00_History.slim the population reduces to n_ind_exp in the last generation to simulate the sampling of the parents for the experiment
n_sample = n_ind_exp                   # Number of individuals to be sampled to construct the c matrix  (This is just because c matrices become awfully large). Typically should be the same as n_ind_exp 

sequence_length = 1e+06                # Just have a single continuous chromosome that is simulated

map_length = ifelse(Sys.info()["nodename"]=="SCE-BIO-C06645", 1.4, (as.numeric(commandArgs(trailingOnly = TRUE)[2])))
map_length_expt = ifelse(Sys.info()["nodename"]=="SCE-BIO-C06645", 1.4, (as.numeric(commandArgs(trailingOnly = TRUE)[3])))

r = map_length/sequence_length             # Recombination rate (per site per generation) during the forward simulation of history
r_expt = map_length_expt/sequence_length   # Recombination rate to be used during during the experiment (Drosophila melanogaster ~ 1.4e-08)
r_msp = r/532                          # Recombination rate for the initial msprime simulation

AtleastOneRecomb = FALSE               # Whether there has to be at least one recombination event

## Mutation rate of non_neutral mutations during the forward simulation of the history
if(Sys.info()["nodename"]=="SCE-BIO-C06645"){mu_list=seq(3e-8, 2e-7, length = nsims)}else{mu_list=seq(commandArgs(trailingOnly = TRUE)[1], commandArgs(trailingOnly = TRUE)[1], length = nsims)}  # If mu is to be varied in order to vary true Vw 

mu_expt = 0                             # Mutation rate during the experiment

# mu_msp and mu_neutral are specified within the loop over sims

##############################
### DFE-related parameters ###
##############################

DFE = "g"                              # DFE can be "g" (gamma) or "n" (normal) 

# If DFE is "g"
shape = 0.3                                     # Shape of the gamma DFE ##### mean = shape*scale
scale_list = seq(0.033, 0.033, length = nsims)  # Vector of Scale of the gamma DFE

# The ratio of beneficial:deleterious mutations 
if(Sys.info()["nodename"]=="SCE-BIO-C06645"){mut_ratio=0}else{mut_ratio = as.numeric(commandArgs(trailingOnly = TRUE)[8])}

# If DFE is "n" need to specify the mean and the variance of the normal distribution
mean_alpha = 0
var_alpha_list = seq(0.00002, 0.0002, length = nsims) # Vector to store variance of normal DFE


###################################################################################################################
##### Create a spreadsheet to store cumulative data across all simulations (only if it doesn't exist already) #####
###################################################################################################################

# Assign a "Set_ID" to each set of simulations 
# The set ID is the name of the system + time at which this script starts running. It stays constant throughout (for all the simulations run in that particular set).

Set_ID = as.character(paste(Sys.info()["nodename"], Sys.time()))
Set_ID = gsub(" ", "_", Set_ID)
Set_ID = gsub(":", "_", Set_ID)
Set_ID = paste(Job_ID, Set_ID, sep="_")

## Only attach command line parameters to the set id if not running on the PC
if(Sys.info()["nodename"]!="SCE-BIO-C06645"){
  Set_ID = paste(Set_ID, commandArgs(trailingOnly = TRUE)[1], commandArgs(trailingOnly = TRUE)[2], commandArgs(trailingOnly = TRUE)[3], commandArgs(trailingOnly = TRUE)[4], commandArgs(trailingOnly = TRUE)[5], commandArgs(trailingOnly = TRUE)[6], commandArgs(trailingOnly = TRUE)[7], commandArgs(trailingOnly = TRUE)[8], sep = "_")
}

if(!file.exists(paste(output_path, "/", Set_ID, "_Data.csv", sep = ""))){
  col_names = as.matrix(t(c("Set_ID","sim", "Time","end_gen", "ngen1", "ngen2", "Ne", "n_ind_exp", "n_cages", "sequence_length", "r_msp", "r", "r_expt", "mu_msp", "mu", "mu_neutral", "shape", "scale", "mut_ratio", "flip_sel_coef")))
  write.table(col_names, file = paste(output_path, "/", Set_ID, "_Data.csv", sep = ""),col.names = FALSE, row.names = FALSE, sep = ",")
}



###########################
### Warnings and checks ###
###########################

if(!DFE%in%c("n", "g")){stop("DFE must be one of 'g', 'n'")}
if(!flip_sel_coef%in%c(0, 1)){stop("flip_sel_coef must be one of 0 or 1")}

# Add a check that stops the sim if nsim > 1 on one of the two clusters (AC3 or Eddie)
# On AC3 and Eddie nsims should always be 1 with parameters varied usin command line arguments.

if(Sys.info()["nodename"]%in%c("bigfoot", "bigshot", "bigbird", "bigyin", "biggar", "bigwig", "c1", "c2", "c3", "c4", "c5", "c6")|grepl("ecdf.ed.ac.uk", Sys.info()["nodename"])){
  if(nsims!=1){stop("nsims must be 1 when running on a cluster")}
}


###########################
##### Run simulations #####
###########################


for (sim in 1:nsims){
  
      # Specify the scale of the the gamma distribution or the variance of the normal distribution of selection coefficients
    
      scale = scale_list[sim] 
      var_alpha = var_alpha_list[sim]
      
      # Specify the mutation rate for the SLiM history simulation
      
      mu = mu_list[sim]
      mu_msp = ifelse(end_gen==2, mu/5320, mu/5320)
      # mu_neutral = ifelse(end_gen==2, mu_msp/3, mu/20)
      mu_neutral = mu_msp/3
      
      message(paste("Simulation", sim, "in progress..."))
    
    
      #####################################################
      ############# Run msprime (neutral burnin) ##########
      #####################################################
      
      message("Running msprime...")
      
      system(paste("python", msprime_burnin_path, Ne, n_ind, sequence_length, r_msp, mu_msp, shape, scale, slim_output_path, mut_ratio, DFE, mean_alpha, sqrt(var_alpha), Set_ID, sim))
      
        
      ###############################################################
      ################# Simulate the history in SLim ################
      ###############################################################
      
      message("Forward simulating the history using SLiM...")
      
      ### This has lot's of command line arguments. Creating separate strings for each argument
      
      arg0 = paste("-d n_ind=", n_ind, sep = "")
      arg1 = paste("-d mu=", mu, sep = "")
      arg2 = paste("-d shape=", shape, sep = "")
      arg3 = paste("-d scale=", scale, sep = "")
      arg4 = paste("-d sequence_length=", sequence_length, sep = "")
      arg5 = paste("-d r=", r, sep = "")
      arg6 = paste("-d ", shQuote(paste("msprime_output_path=","'", slim_output_path, "'", sep = "")), sep = "") # Path where the msprime burnin is stored (it's same as slim_output_path)
      arg7 = paste("-d ", shQuote(paste("slim_output_path=","'", slim_output_path, "'", sep = "")), sep = "") 
      arg8 = paste("-d end_gen=", end_gen, sep = "")
      arg9 = paste("-d output_freq=", output_freq, sep = "")
      arg10 = paste("-d mut_ratio=", mut_ratio, sep = "")
      arg11 = paste("-d n_ind_exp=", n_ind_exp, sep = "") # The population size is to be reduced to n_ind_exp in generation end_gen (to simulate sampling of the parents for the experiment)
      arg12 = paste("-d ", shQuote(paste("DFE=", "'", DFE, "'", sep = "")), sep = "")
      arg13 = paste("-d mean_alpha=", mean_alpha, sep = "")
      arg14 = paste("-d var_alpha=", var_alpha, sep = "")
      arg15 = paste("-d ", shQuote(paste("Set_ID=", "'", Set_ID, "'", sep = "")), sep = "")
      arg16 = paste("-d simulation=", sim, sep = "")
      arg17 = paste("-d flip_sel_coef=", flip_sel_coef, sep = "")
      
      system(paste("slim", arg0, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9, arg10, arg11, arg12, arg13, arg14, arg15, arg16, arg17, slim_history_path))
      
      ###################################################################
      ############### Add neutral mutations using msprime ###############
      ###################################################################
      
      message("Adding neutral mutations...")
      system(paste("python", msprime_add_neutral_path, slim_output_path, mu_neutral, Set_ID, sim))
      
      ######################################################################
      ##### Randomly swap selection coefficients (if flip_sel_coef==T) #####
      ######################################################################
      
      # In this SLiM script, we first read the .trees file generated after adding neutral mutations, i.e. slim_output_path + "/" + Set_ID + "_sim" + simulation + "_output_history_with_neutral.trees".
      # We then randomly swap signs of selection coefficients *** if flip_sel_coef==T ***, and save two output files
      # 1. a .txt file (slim_output_path + "/" + Set_ID + "_sim" + simulation + "_output_parents.txt") = This file is for downstream snalyses, it is not read by 4_Experiment.slim script
      # 2. a .trees file (slim_output_path + "/" + Set_ID + "_sim" + simulation + "_output_history_with_neutral.trees", metadata = sel_coef)
      # By default the .trees file does not record changes to the selection coefficients. Therefore while saving the .trees file we also need to pass as metadata a dictionary containing the mutation ids and selection coeficients
      # All in just one tick
      # This script does not run the simulation
      # It's just a hack to read the .trees output, swap selection coefficients (if required), and then store the output as a .txt file
      # The flip_sel_coef==T check happens within the script
      # Irrespective of whether flip_sel_coef==T, the script is run, minimally to convert the output from .trees to .txt
      
      if(flip_sel_coef==1){message("Randomly flipping selection coefficients and Coverting .trees output in the parents' generation to .txt for analyses...")}else{message("Coverting .trees output in the parents' genreation to .txt for analyses...")}
      
      ### This has lot's of command line arguments. Creating separate strings for each argument
      # Strictly speaking the parameters regarding DFE shouldn't be necessary, but keeping them on just in case
      
      flip_arg1 = paste("-d shape=", shape, sep = "")
      flip_arg2 = paste("-d scale=", scale, sep = "")
      flip_arg3 = paste("-d sequence_length=", sequence_length, sep = "")
      flip_arg4 = paste("-d ", shQuote(paste("slim_output_path=","'", slim_output_path, "'", sep = "")), sep = "") 
      flip_arg5 = paste("-d mut_ratio=", mut_ratio, sep = "")
      flip_arg6 = paste("-d ", shQuote(paste("DFE=", "'", DFE, "'", sep = "")), sep = "")
      flip_arg7 = paste("-d mean_alpha=", mean_alpha, sep = "")
      flip_arg8 = paste("-d var_alpha=", var_alpha, sep = "")
      flip_arg9 = paste("-d ", shQuote(paste("Set_ID=", "'", Set_ID, "'", sep = "")), sep = "")
      flip_arg10 = paste("-d simulation=", sim, sep = "")
      flip_arg11 = paste("-d flip_sel_coef=", flip_sel_coef, sep = "")
      
      system(paste("slim", flip_arg1, flip_arg2, flip_arg3, flip_arg4, flip_arg5, flip_arg6, flip_arg7, flip_arg8, flip_arg9, flip_arg10, flip_arg11, slim_flip_sel_coef_path))
      
      ###################################################
      ######### Simulate the experiment in SLiM #########
      ###################################################
      
      message("Forward simulating the experiment using SLiM...")
      
      ### Loop over the cage replicates
      
      for (cage in (1:n_cages)){
        
        message(paste("Cage", cage, "of simulation", sim, "in progress..."))
        
        
        ### This has lot's of command line arguments. Creating separate strings for each argument
        
        expt_arg1 = paste("-d mu_expt=", mu_expt, sep = "")
        expt_arg2 = paste("-d shape=", shape, sep = "")
        expt_arg3 = paste("-d scale=", scale, sep = "")
        expt_arg4 = paste("-d sequence_length=", sequence_length, sep = "")
        expt_arg5 = paste("-d r=", r_expt, sep = "")
        expt_arg6 = paste("-d ", shQuote(paste("slim_output_path=","'", slim_output_path, "'", sep = "")), sep = "") 
        expt_arg7 = paste("-d mut_ratio=", mut_ratio, sep = "")
        expt_arg8 = paste("-d end_gen=", end_gen, sep = "")
        expt_arg9 = paste("-d ", shQuote(paste("DFE=", "'", DFE, "'", sep = "")), sep = "")
        expt_arg10 = paste("-d mean_alpha=", mean_alpha, sep = "")
        expt_arg11 = paste("-d var_alpha=", var_alpha, sep = "")
        expt_arg12a = paste("-d ngen1=", ngen1, sep = "")
        expt_arg12b = paste("-d ngen2=", ngen2, sep = "")
        expt_arg13 = paste("-d ", shQuote(paste("Set_ID=", "'", Set_ID, "'", sep = "")), sep = "")
        expt_arg14 = paste("-d simulation=", sim, sep = "")
        expt_arg15 = paste("-d cage=", cage, sep = "")
        expt_arg16 = paste("-d flip_sel_coef=", flip_sel_coef, sep = "")
        
        system(paste("slim", expt_arg1, expt_arg2, expt_arg3, expt_arg4, expt_arg5, expt_arg6, expt_arg7, expt_arg8, expt_arg9, expt_arg10, expt_arg11, expt_arg12a, expt_arg12b, expt_arg13, expt_arg14, expt_arg15, expt_arg16, slim_expt_path))
        
        ### Trim the SLim outputs for the experiment to just include the information on mutations to make them smaller
        
        if (trim_exp_files==TRUE){
          message("Trimming output files for the experiment...")
          for (gen in (end_gen + 1):(end_gen + 1 + ngen_expt)){
            filename = paste(slim_output_path, "/", Set_ID, "_sim", sim, "_cage", cage, "_output_experiment_", gen, ".txt", sep = "")
            temp_filename = paste(slim_output_path, "/", Set_ID, "_tempfile.txt", sep = "")
            system(paste("sed -n '/Mutations:/, /Individuals:/p'", filename, ">", temp_filename, sep = " ")) # Save the trimmed text to temp_file
            system(paste("mv", temp_filename, filename))                                                     # Rename the temp_file as the original file 
            
          }
        }
        
      }
      
      ########################################################
      ######### Save simulation data in a spreadsheet ########
      ########################################################
      
      # If analyses are not to be performed record simulation data here
      
      if(record == TRUE){
        dat = read.csv(paste(output_path, "/", Set_ID, "_Data.csv", sep = ""), header=FALSE)
        dat = rbind(dat, c(Set_ID, sim, as.character(Sys.time()), end_gen, ngen1, ngen2, Ne, n_ind_exp, n_cages, sequence_length, r_msp, r, r_expt, mu_msp, mu, mu_neutral, shape, scale, mut_ratio, flip_sel_coef))
        write.table(dat, file = paste(output_path, "/", Set_ID, "_Data.csv", sep = ""),col.names = FALSE, row.names = FALSE, sep = ",")
      }
      
      ###########################
      ### Remove .trees files ###
      ###########################
      
      if(del_files){
        message("Deleting .trees files...")
        neutral_burnin_file = paste(slim_output_path, "/", Set_ID, "_sim", sim, "_neutral_burnin.trees", sep = "")
        output_history_file = paste(slim_output_path, "/", Set_ID, "_sim", sim, "_output_history.trees", sep = "")
        output_history_neutral_file = paste(slim_output_path, "/", Set_ID, "_sim", sim, "_output_history_with_neutral.trees", sep = "")
        system(paste("rm -rf", neutral_burnin_file, output_history_file, output_history_neutral_file))
      }
      
      #########################################
      ### Compress .txt and/or .trees files ###
      #########################################
      
      if(compress_files){
        message("Compressing files...")
        system(paste("gzip ", slim_output_path, "/", Set_ID, "*.txt", sep = ""))
        system(paste("gzip ", slim_output_path, "/", Set_ID, "*.trees", sep = ""))
      }
      
      message(paste("Simulation", sim, "of set", Set_ID, "completed...")) 
       
}


